### ISSUE TITLE
Int4 Support

### ISSUE THREAD

**Issue by fmac2000** [OpenNMT/CTranslate2#1104]
Hello Authors,

I apologise for asking questions unrelated to an issue with the repo however, would you consider support a newer paradigm I came across whilst reading a recent [paper](https://www.researchgate.net/publication/367557918_Understanding_INT4_Quantization_for_Transformer_Models_Latency_Speedup_Composability_and_Failure_Cases)?

It looks incredibly promising and rather well written I must say, especially when considering the performance of such a precision.
Is there anyone on the team able to give this a shot?

**Comment by guillaumekln** [comment]
Hello,

Thank you for sharing this paper!

At this time I don't plan on integrating INT4 which would require using [CUTLASS](https://github.com/NVIDIA/cutlass) to define custom kernels. We are currently using cuBLAS for matrix multiplication.

**Comment by jncraton** [comment]
Would it be reasonable to implement this as a CPU-only optimization? [GGML](https://github.com/ggerganov/ggml) does this, but I'm not sure if that approach makes sense here or not.

**Comment by Matthieu-Tinycoaching** [comment]
Hi,

Would be great to have the possibility to integrate int4 quantization regarding the very interesting results in terms of performance and inference!

**Comment by bil-ash** [comment]
@guillaumekln Yes, 4bit quantization (on cpu) is a very much required feature. Any plans of taking this up?

**Comment by bil-ash** [comment]
Or maybe @ebraraktas can go one step further and implement 2bit and 3bit quantization using by taking clues from https://github.com/intel/neural-speed/pull/178

**Comment by thiswillbeyourgithub** [comment]
I am a user of CTranslate2 because I serve whisper via [speaches](https://github.com/speaches-ai/speaches). Using int8 has been incredibly nice, but I'm very interested in int4.

Just wanted to voice that there is still interest for that feature in 2025.
